p4include/psa.p4

```diff
@@ -0,0 +1,672 @@
+
+  // BufferingQueueingEngine(); /// No constructor. BQE is instantiated
+                                /// by the architecture.
+
+  /// Unicast operation: sends packet to port
+  ///
+  /// Targets may implement this operation by setting the appropriate
+  /// intrinsic metadata or through some other mechanism of
+  /// configuring the BQE.
+  ///
+  /// @param port The output port. If the port is PORT_CPU the packet
+  ///             will be sent to CPU
+  void send_to_port (in PortId_t port);
+
+  /// Drop operation: do not forward the packet
+  ///
+  /// The PSA implements drop as an operation in the BQE. While the
+  /// drop operation can be invoked anywhere in the ingress pipeline,
+  /// the semantics supported by the PSA is that the drop will be at
+  /// the end of the pipeline (ingress or egress).
+  void drop      ();
+
+  /// Truncate operation
+  ///
+  /// Truncate the outgoing packet to the specified length
+  ///
+  /// @param length packet length
+  void truncate(in bit<32> length);
+
+  /*
+  @ControlPlaneAPI
+  {
+  // TBD
+  }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/*
+ * Hashes
+ */
+
+/// Supported hash algorithms
+enum HashAlgorithm {
+  crc32,
+  crc32_custom,
+  crc16,
+  crc16_custom,
+  random,           /// are random hash algorithms useful?
+  identity
+}
+
+/// Hash function
+///
+/// @param Algo The algorithm to use for computation (@see HashAlgorithm).
+/// @param O    The type of the return value of the hash.
+///
+/// Example usage:
+/// parser P() {
+///   Hash<crc16, bit<56>> h();
+///   bit<56> hash_value = h.getHash(16, buffer, 100);
+///   ...
+/// }
+extern Hash<Algo, O> {
+  /// Constructor
+  Hash();
+
+  /// compute the hash for data
+  O getHash<T, D, M>(in T base, in D data, in M max);
+
+  /*
+  @ControlPlaneAPI
+  { }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/// Checksum computation
+///
+/// @param W    The width of the checksum
+extern Checksum<W> {
+  Checksum(HashAlgorithm hash);          /// constructor
+  void clear();              /// prepare unit for computation
+  void update<T>(in T data); /// add data to checksum
+  void remove<T>(in T data); /// remove data from existing checksum
+  W    get();      	     /// get the checksum for data added since last clear
+
+  /*
+  @ControlPlaneAPI
+  { }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Counters
+ *
+ * Counters are a simple mechanism for keeping statistics about the
+ * packets that trigger a table in a Match Action unit.
+ *
+ * Direct counters fire when the count method is ivoked in an action, and
+ * have an instance for each entry in the table.
+ */
+
+/// Counter types
+enum CounterType_t {
+    packets,
+    bytes,
+    packets_and_bytes
+}
+
+/// Counter: RFC-XXXX
+extern Counter<W, S> {
+  Counter(S n_counters, W size_in_bits, CounterType_t counter_type);
```

Comment:
We have to be careful here. For example if this is packets_and_bytes, then the width for the packet and byte counter (if less than 64 bits) is typically different and thus might require two separate arguments.